templatekit/repo.py

```diff
@@ -0,0 +1,277 @@
+            except (OSError, ValueError) as err:
+                # Not a template directory
+                message = ('Found project_template directory {0!r} but it is '
+                           'not a recognizable template. {1!s}')
+                logging.warning(message.format(template_dir, err))
+                continue
+            yield template
+
+    def _list_directory_items(self, dirname):
+        fs_items = os.listdir(dirname)
+        fs_items.sort()
+        fs_items = [os.path.join(dirname, item) for item in fs_items]
+        return [fs_item for fs_item in fs_items if os.path.isdir(fs_item)]
+
+
+class BaseTemplate(object):
+    """Template (file or project) in the templates repo.
+
+    Parameters
+    ----------
+    path : `str`
+        Path of the template's directory.
+
+    Raises
+    ------
+    OSError
+        Raised if ``path`` is not a directory.
+    ValueError
+        Raised if ``path`` is a directory that does not contain a recognizable
+        template.
+    """
+
+    def __init__(self, path):
+        super().__init__()
+        self._log = logging.getLogger(__name__)
+        self.path = os.path.abspath(path)
+        if not os.path.isdir(self.path):
+            message = 'File template directory {} not found.'.format(self.path)
+            raise OSError(message)
+
+        self._validate_template_dir()
+
+    def _validate_template_dir(self):
+        """Run a quick set of checks that this is in fact a template
+        repository, with a cookiecutter.json directory, etc.
+        """
+        if not os.path.isfile(self.cookiecutter_json_path):
+            message = 'cookiecutter.json not found in {}'.format(self.path)
+            raise ValueError(message)
+
+    def __str__(self):
+        return '{0!s}({1!r})'.format(self.__class__.__name__, self.name)
+
+    def __repr__(self):
+        return '{0!s}({1!r})'.format(self.__class__.__name__, self.name)
+
+    @property
+    def name(self):
+        """Name of the template (`str`).
+        """
+        return os.path.split(self.path)[-1]
+
+    @property
+    def cookiecutter_json_path(self):
+        """Path of the cookiecutter.json file (`str`).
+        """
+        return os.path.join(self.path, 'cookiecutter.json')
+
+
+class FileTemplate(BaseTemplate):
+    """File template.
+
+    Parameters
+    ----------
+    path : `str`
+        Path of the template's directory.
+
+    Raises
+    ------
+    OSError
+        Raised if ``path`` is not a directory.
+    ValueError
+        Raised if ``path`` is a directory that does not contain a recognizable
+        template.
+    """
+
+    @property
+    def source_path(self):
+        """Path to the template source file (a .jinja extension) (`str`).
+        """
+        items = os.listdir(self.path)
+        for item in items:
+            if os.path.splitext(item)[-1] == '.jinja':
+                return os.path.join(self.path, item)
+
+
+class ProjectTemplate(BaseTemplate):
```

Comment:
Since this adds nothing, what do you think about renaming `BaseTemplate` to `ProjectTemplate` and have `FileTemplate` inherit from `ProjectTemplate`?  Is the main concern that you will eventually add attributes to `ProjectTemplate` that you don't want showing up in `FileTemplate`?